protocol/proto/types.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"net"
+
+	"github.com/m3db/m3x/pool"
+)
+
+// Marshaler can be marshaled.
+type Marshaler interface {
+	Size() int
+	MarshalTo(data []byte) (int, error)
+}
+
+// Unmarshaler can be unmarshaled from bytes.
+type Unmarshaler interface {
+	Unmarshal(data []byte) error
+}
+
+// Encoder encodes the marshaler.
+type Encoder interface {
+	Encode(m Marshaler) error
+}
+
+// Decoder decodes into an unmarshaler.
+type Decoder interface {
+	Decode(m Unmarshaler) error
+}
+
+// EncodeDecoder can encode and decode.
+type EncodeDecoder interface {
+	Encoder
+	Decoder
+
+	// Close closes the EncodeDecoder.
+	Close()
+
+	// Reset resets the EncodeDecoder.
+	Reset(conn net.Conn)
```

Comment:
Do you need a connection? Can we use `io.ReadWriter`? Seems a bit weird to have a connection in the encoder decoder interface.

Reply:
Yeah can do that